lookout/style/typos/research/nn_prediction.py

```diff
@@ -0,0 +1,146 @@
+import sys
+from typing import Iterable, Sequence, Tuple
+
+from gensim.models.fasttext import FastText
+try:
+    import keras
+except ImportError:
+    sys.exit("TensorFlow is required for this analyzer")
+from keras.layers import Dense, BatchNormalization, Activation
+from keras.models import Sequential
+import numpy
+import pandas
+
+from lookout.style.typos.utils import extract_embeddings_from_fasttext, \
+    CORRECT_TOKEN_COLUMN, TYPO_COLUMN
+
+
+def get_features(fasttext: FastText, typos: Sequence[str]) -> numpy.ndarray:
+    return numpy.concatenate((extract_embeddings_from_fasttext(fasttext, typos),
+                             numpy.ones((len(typos), 1))), axis=1)
+
+
+def get_target(fasttext: FastText, identifiers: Iterable[str]) -> numpy.ndarray:
+    return extract_embeddings_from_fasttext(fasttext, identifiers)
+
+
+def generator(features: numpy.ndarray, target: numpy.ndarray, batch_size: numpy.ndarray
+              ) -> Tuple[numpy.ndarray, numpy.ndarray]:
+    """
+    Pumps the data for keras.Model.fit_generator()
+
+    :param features: Inputs.
+    :param target: Labels.
+    :param batch_size: Batch size.
+    :return: Another batch for fit_generator().
+    """
+
+    while True:
+        indices = numpy.random.randint(features.shape[0], size=batch_size)
+        batch_features = features[indices]
+        batch_target = target[indices]
+        yield batch_features, batch_target
+
+
+def create_model(num_neurons: int, input_len: int, output_len: int) -> keras.models.Sequential:
+    """
+    Builds the fully-connected NN.
+
+    :param num_neurons: Number of neurons in each hidden layer.
+    :param input_len: Input size.
+    :param output_len: Output size.
+    :return: Built model.
+    """
+    model = Sequential()
+    model.add(Dense(units=num_neurons, input_dim=input_len))
+    model.add(BatchNormalization())
+    model.add(Activation("relu"))
+    model.add(Dense(units=num_neurons))
+    model.add(BatchNormalization())
+    model.add(Activation("relu"))
+    model.add(Dense(units=output_len))
+    return model
```

Comment:
BN is normally used after ReLU (https://www.reddit.com/r/MachineLearning/comments/67gonq/d_batch_normalization_before_or_after_relu/)

Reply:
https://github.com/ducha-aiki/caffenet-benchmark/blob/master/batchnorm.md

Yep, the numbers do not lie.